deepmd/train/trainer.py

```diff
@@ -597,69 +654,43 @@ def _build_network(self, data, suffix=""):
             reuse=False,
         )
 
-        if not self.multi_task_mode:
-            self.l2_l, self.l2_more = self.loss.build(
-                self.learning_rate,
-                self.place_holders["natoms_vec"],
-                self.model_pred,
-                self.place_holders,
-                suffix="test",
-            )
-
-            if self.mixed_prec is not None:
-                self.l2_l = tf.cast(
-                    self.l2_l, get_precision(self.mixed_prec["output_prec"])
-                )
-        else:
-            self.l2_l, self.l2_more = {}, {}
-            for fitting_key in self.fitting_type_dict:
-                self.l2_l[fitting_key], self.l2_more[fitting_key] = self.loss_dict[
-                    fitting_key
-                ].build(
-                    self.learning_rate,
-                    self.place_holders["natoms_vec"],
-                    self.model_pred[fitting_key],
-                    self.place_holders,
-                    suffix=fitting_key,
-                )
-                if self.mixed_prec is not None:
-                    self.l2_l[fitting_key] = tf.cast(
-                        self.l2_l[fitting_key],
-                        get_precision(self.mixed_prec["output_prec"]),
-                    )
+        self.l2_l, self.l2_more = self._build_optimizer()
 
         log.info("built network")
 
     def _build_training(self):
         trainable_variables = tf.trainable_variables()
-        if self.run_opt.is_distrib:
-            if self.scale_lr_coef > 1.0:
-                log.info("Scale learning rate by coef: %f", self.scale_lr_coef)
-                optimizer = tf.train.AdamOptimizer(
-                    self.learning_rate * self.scale_lr_coef
-                )
-            else:
-                optimizer = tf.train.AdamOptimizer(self.learning_rate)
-            optimizer = self.run_opt._HVD.DistributedOptimizer(optimizer)
-        else:
-            optimizer = tf.train.AdamOptimizer(learning_rate=self.learning_rate)
-        if self.mixed_prec is not None:
-            _TF_VERSION = Version(TF_VERSION)
-            # check the TF_VERSION, when TF < 1.12, mixed precision is not allowed
-            if _TF_VERSION < Version("1.14.0"):
-                raise RuntimeError(
-                    "TensorFlow version %s is not compatible with the mixed precision setting. Please consider upgrading your TF version!"
-                    % TF_VERSION
-                )
-            elif _TF_VERSION < Version("2.4.0"):
-                optimizer = tf.train.experimental.enable_mixed_precision_graph_rewrite(
-                    optimizer
-                )
-            else:
-                optimizer = tf.mixed_precision.enable_mixed_precision_graph_rewrite(
-                    optimizer
-                )
         if not self.multi_task_mode:
```

Comment:
Do not write large if-else condition. try to pack optimizer-making codes in functions and call the functions in conditions.